src/inmanta/agent/handler.py

```diff
@@ -961,6 +977,109 @@ def execute(self, ctx: HandlerContext, resource: TPurgeableResource, dry_run: bo
         super().execute(ctx, resource, dry_run)
 
 
+@stable_api
+class DiscoveryHandler(HandlerABC, Generic[DRT, URT]):
+    # The DiscoveryHandler is generic in both the handler's Discovery Resource type (DRT)
+    # and the Unmanaged Resource type (URT) it reports to the server. The second has to be serializable.
+
+    # This class deploys instances of DRT and reports URT to the server.
+
+    def __init__(self, agent: "inmanta.agent.agent.AgentInstance") -> None:
+        self._agent = agent
+        self._ioloop = agent._io_loop
+
+    @abstractmethod
+    def discover_resources(self, ctx: HandlerContext, discovery_resource: DRT) -> abc.Mapping[ResourceIdStr, URT]:
+        raise NotImplementedError
+
+    def report_discovered_resources(self, ctx: HandlerContext, resource: DRT) -> None:
+        """ """
+
+        def _call_discovered_resource_create_batch() -> typing.Awaitable[Result]:
+            return self.get_client().discovered_resource_create_batch(
+                tid=self._agent.environment, discovered_resources=discovered_resources
```

Comment:
Just a style preference but I would always pass `discovered_resources` as a parameter rather than counting on an implicit closure of something that's defined below it.